rfc/rfc-0000-k8s-events.md

```diff
@@ -0,0 +1,170 @@
+[in this rfc](https://github.com/vmware-tanzu/cartographer/pull/738). Although there is no art to tackle sharing
+per-resource 'Submitted' status (I will raise a separate RFC), it's something that can be handled entirely using the
+status in a workload (ie, it does not require the use of events)
+
+With different initiatives (in RFCs for workload status improvements) the developer's view of information is becoming
+quite complete, but a key issue is the absence of awareness that "something is happening", especially after they've
+committed new code. With event's we can provide information that show's activity and causality not present in the
+current snapshot of state.
+
+We add events to the Owners (Workload/Deliverable) to enable debug and reasoning about Cartographer's behavior.
+
+## Operator View
+
+The situation is worse for users managing the supply chains (Devops/Ops). By looking at a supply chain, there is very
+little information about how supply chains are behaving, or if any issues are occurring for developers:
+
+![Blueprint](./rfc-k8s-events/blueprint.jpg)
+
+Although these users can access the logs, we can provide them a more concise view of how supply chains are behaving
+on-cluster.
+
+We add events to the Blueprints (Supply Chain/Delivery) to enable debug and reasoning about Cartographer's behavior.
+
+# How it Works
+
+[how-it-works]: #how-it-works
+
+To avoid over-communicating, and making the view of event's overwhelming and dilute for user's we will start with two
+key guiding principles:
+
+1. > Let it be something that isn't satisfactorily presented in other user views (eg: `k get workload`).
+2. > Let the event list be as meaningful as is necessary, and no more.
+
+The cartographer controller could emit the following events:
+
+| Reason | Message Format | Description | involvedObject |
+| --- | --- | --- | --- |
+| ResourceExternalSpecChange | an external actor changed the spec of `<group>.<kind>/<name>` | Our cache of the spec, and the spec we just generated match, but the API server has a different one. A lot of these is evidence of thrashing with external resources | Owner |
+| ResourceExternalSpecChange | an external actor changed the spec of resource: `<resource name>` for object `<group>.<kind>` | Our cache of the spec, and the spec we just generated match, but the API server has a different one. A lot of these is evidence of thrashing with external resources | Blueprint |
+| ResourceGetError | `<resource name>` could not retrieve `<group>.<kind>/<name>` due to error: `<error message>` | Loading a resource is failing due to a client.get issue (missing is not an error) | Owner |
+| ResourceGetError | `<resource name>` could not retrieve `<group>.<kind>` due to error: `<error message>` | Loading a resources is failing due to a client.get issue (missing is not an error) - this could be spammy | Blueprint |
```

Comment:
I'm curious about this one and why we'd want to expose this event on the blueprint? I'd imagine this would be something like a permissions issue (unless I'm thinking about this wrong). If that is the case, is the thinking that the operator would want to know about this event because they're likely the one supplying the service account the owner is using?

Reply:
I realize I'm assuming that the ResourceGetError is that cartographer can't get the applied object from the API server, but maybe that's wrong. Is this supposed to represent that cartographer could not get the template the resource is referring to? If so, what kind of error would this be that would be different from ResourceDoesNotExist?

Reply:
No, I was thinking of the most highly visible and problematic partitioning errors are API errors. There are probably far better ways of discovering this about your cluster so the merit here is pretty weak.